### ISSUE TITLE
Add option to write diagnostics from balance operator

### ISSUE THREAD

**Issue by travissluka** [JCSDA-internal/soca#1106]
## Description

Add an option to the balance operator linvarchange to write out the Jacobians and evaluate how they look

(See what was done in the old BkgErrGodas lin variable change
https://github.com/JCSDA-internal/soca/blob/7f2ddb61bc86796c83dfcd4801c91bffd829ffb4/src/soca/LinearVariableChange/BkgErrGodas/soca_bkgerrgodas_mod.F90#L123 )

- [ ] add option to save the T/S and SSH/TS  Jacobians
- [ ] save example diagnostics for several different months of the year (it would be good to capture NH summer/winter)
- [ ] take a look and see if things look reasonable

**Comment by Dooruk** [comment]
Is there an example for writing netcdf files in SOCA without defining the `type` as `soca_fields`? Because I can't use `write_file` for jacobians otherwise. I couldn't find via code search.

For instance, in your bkgerr pointer:
```fortran
  type(soca_fields)                 :: std_bkgerr
```

**Comment by travissluka** [comment]
> Is there an example for writing netcdf files in SOCA without defining the `type` as `soca_fields`? Because I can't use `write_file` for jacobians otherwise. I couldn't find via code search.
> 
> For instance, in your bkgerr pointer:
> 
>   type(soca_fields)                 :: std_bkgerr

ah yes, one of the downsides of our soca IO, this is one of my motivations for refactoring things into C++/native-atlas fieldsets, it's easy to dump out any arbitrary atlas fieldset on the C++ side. You could 
 1) use netcdf calls to create a file yourself  (chatgpt might be able to easily to this for you!) or 
 2) still use soca_fields so that you can use the soca io, if you want new variable names, you can add whatever you want to the field_metadata yaml file

**Comment by Dooruk** [comment]
Ok thanks! I guess it depends on if you want this as a more permanent feature to be used among other balance, so I could even add it as a method/module, or is it like a one off time while we fix things. I will try both ways.

**Comment by travissluka** [comment]
> Ok thanks! I guess it depends on if you want this as a more permanent feature to be used among other balance, so I could even add it as a method/module, or is it like a one off time while we fix things. I will try both ways.

yeah, I'd say do whatever is fastest for you. It's in the plans to refactor all this into a C++ saber outer block

**Comment by Dooruk** [comment]
So, I want to go with the `soca_fields` route to have dimensions in the `jac_out` output file. When I add this part i get the following error.

```fortran
  type(soca_field), pointer :: tocn, socn, hocn, cice, mld, layer_depth
  type(soca_fields) :: jac_out

.
.
.
    ! optionally write out Kst Jacobian
    if( f_conf%has("jacobian_output")) then
      call f_conf%get_or_die("jacobian_output.filename", str)
      call self%jac_out%copy(self%kst%jacobian)
      ! call self%kst%jacobian%write_file(str)
    end if

```

`This is not a component name that is defined in the encompassing structure.   [JAC_OUT] call self%jac_out%initialize()`

Which is confusing because I'm solely replicating what was done for `std_bkgerr`. I don't see any specific changes for `std_bkgerr` to work in the past SOCA versions, what am I missing?

Sorry, I'm not a JEDI Academy attendee so bear with my simple questions. I'm one level below Padawan if that exists 😄 Happy to have brief chats over these type of minor issues if you have time.

**Comment by Dooruk** [comment]
Ok, I was able to use `soca_fields` properly to have an optional config test writing other existing `soca_fields`.

I would like to write jacobians with coordinate metadata, and copying fields within SOCA has some restrictions.

> still use soca_fields so that you can use the soca io, if you want new variable names, you can add whatever you want to the field_metadata yaml file

Does this mean i can just create a new variable called jacobian within the ocean group?

Otherwise I can write jacobians to a netcdf file via fortran modules and use coordinates from `soca_gridspec.nc` or something.

**Comment by travissluka** [comment]
> Does this mean i can just create a new variable called jacobian within the ocean group?
> Otherwise I can write jacobians to a netcdf file via fortran modules and use coordinates from `soca_gridspec.nc` or something.

Yes, you can add any arbitrary fields that you want in the https://github.com/JCSDA-internal/soca/blob/develop/test/Data/fields_metadata.yml file.
so, something like  this should work
```yaml
- name: jacobian_tssh
  levels: full_ocn
  io file: ocn
  io name: jacobian_tssh
```
then when you create the `soca_fields` make sure `jacobian_tssh` is in the variable list when it's created. But if that's giving you trouble, then yes you can just use direct netcdf calls to write out.

**Comment by Dooruk** [comment]
I played around with the kst parameters and different differentual numerical schemes but I get a somewhat noisy Kst jacobian. I would think we could use some smoothing. Should I wait for balance part to move to C++? This is the transect for lat=0.

![Image](https://github.com/user-attachments/assets/bfdedf8e-d14b-4372-8f82-2b40a8c588ce)
